### DISCUSSION TITLE
Tic Tac Toe contract

### DISCUSSION THREAD

**Discussion by Skjaldbaka17** [stellar/sorobanathon#35] [General]
Hi 🐝,
 
I made a TicTacToe contract with an accompanying tutorial published on [mirror](https://mirror.xyz/0xfcd2E272798b5d9A487c584285E8F22f5472cbB8/PoOA5ZmKakq2eNEsLRVtJCRo9Bzb0CIaEgIYwhsrIOA). Hope some of youse find it useful. + All criticism welcome.🐢

[repo](https://github.com/Skjaldbaka17/tic-tac-toe)

**Comment by leighmcculloch** [comment]
Hi @Skjaldbaka17, very nice!

A couple small things I recommend checking out:

- I noticed that the contract only supports invokes from accounts, and not other contracts. I think the contract could ignore that distinction and use an `Address` instead of an `AccountId` to reference players. Contracts can be smart multi-sig wallets, or other forms of account abstractions, so it's helpful for interoperability not to distinguish between the two. Plus, removing the distinction is less logic for the contract to need to worry about.

- Check out the errors example for how to return errors and `panic_error!`: https://soroban.stellar.org/docs/examples/errors. I recommend trying that out instead of using `panic!` and `assert!` because panic messages will be omitted once deployed. (Note that there's no `assert!` that supports error codes yet, but it is coming: https://github.com/stellar/rs-soroban-sdk/pull/749.)

**Reply by Skjaldbaka17** [reply]
Makes sense. Thank you for your feedback🐼 

I've updated the [repo](https://github.com/Skjaldbaka17/tic-tac-toe) and the [article / tutorial](https://mirror.xyz/0xfcd2E272798b5d9A487c584285E8F22f5472cbB8/PoOA5ZmKakq2eNEsLRVtJCRo9Bzb0CIaEgIYwhsrIOA) according to your recommendations:

1. Allow both accounts and contracts to use the functions (removed the accountId check, and swapped accountId for Address in structs).
2. Added errors like the example suggests (using panic_err!)
3. Use **if** instead of **assert**. Look forward to **assert_error** support 🐢 

Let me know if the fixes make sense👶 + If there is anything else you recommend me looking at I'm all ears🦥

**Comment by achallagundla** [comment]
Heya @Skjaldbaka17  ,
Congratulations! Your contribution to Sorobanathon: First Light has been accepted! 🎉
Your submission may qualify for a reward of 1500 XLM! One last step: to receive that reward, please fill out this [form](https://stellarform.typeform.com/Sorobanathon) completely and accurately.

Great piece BTW! Can't wait to see you keep tinkering with Soroban and documenting your finding!

Please let us know if you have any questions. (You can also message me on discord @ anuxhya)

All the best,
Anuhya
